### ISSUE TITLE
Missing requirement in Makefile.PL for `Term::Detect::Software`

### ISSUE THREAD

**Issue by jackdeguest** [perlancar/perl-Color-ANSI-Util#4]
When running `JSON::Color`, it uses `Color::ANSI::Util`, which in turns relies on `Term::Detect::Software`, but is not declared as a dependency in the Makefile.PL, and thus is not installed automatically by CPAN installers such as `cpanm`.

The error produced is:

```
Can't locate Term/Detect/Software.pm in @INC (you may need to install the Term::Detect::Software module) (@INC contains: /home/ai/lib /home/ai/etc /mnt/jack/ai/lib /usr/lib /mnt/jack/ai/etc /etc /usr/local/lib/perl5/site_perl/5.36.1/aarch64-linux-thread-multi /usr/local/lib/perl5/site_perl/5.36.1 /usr/local/lib/perl5/5.36.1/aarch64-linux-thread-multi /usr/local/lib/perl5/5.36.1 /usr/local/lib/perl5/site_perl/5.36.0/aarch64-linux-thread-multi /usr/local/lib/perl5/site_perl/5.36.0 /usr/local/lib/perl5/site_perl) at /usr/local/lib/perl5/site_perl/5.36.1/Color/ANSI/Util.pm line 430.
```

The stack trace is, bearing in mind that `ai.pl` is my script name:

```
    	main::__ANON__[./bin/ai.pl:125]('Can\'t locate Term/Detect/Software.pm in @INC (you may need to install the Term::Detect::Software module) (@INC contains: /home/ai/lib /home/ai/etc /mnt/jack/ai/lib /usr/lib /mnt/jack/ai/etc /etc /usr/local/lib/perl5/site_perl/5.36.1/aarch64-linux-thread-multi /usr/local/lib/perl5/site_perl/5.36.1 /usr/local/lib/perl5/5.36.1/aarch64-linux-thread-multi /usr/local/lib/perl5/5.36.1 /usr/local/lib/perl5/site_perl/5.36.0/aarch64-linux-thread-multi /usr/local/lib/perl5/site_perl/5.36.0 /usr/local/lib/perl5/site_perl) at /usr/local/lib/perl5/site_perl/5.36.1/Color/ANSI/Util.pm line 430.^J') called at /usr/local/lib/perl5/site_perl/5.36.1/Color/ANSI/Util.pm line 430
    	eval {...} at /usr/local/lib/perl5/site_perl/5.36.1/Color/ANSI/Util.pm line 430
    	Color::ANSI::Util::_color_depth at /usr/local/lib/perl5/site_perl/5.36.1/Color/ANSI/Util.pm line 533
    	Color::ANSI::Util::ansi_reset(1) called at /usr/local/lib/perl5/site_perl/5.36.1/JSON/Color.pm line 35
    	JSON::Color::_string('code', 'HASH(0xaaaae1668220)') called at /usr/local/lib/perl5/site_perl/5.36.1/JSON/Color.pm line 153
    	JSON::Color::_hash('HASH(0xaaaae1660610)', 'HASH(0xaaaae17fc408)') called at /usr/local/lib/perl5/site_perl/5.36.1/JSON/Color.pm line 171
    	JSON::Color::_encode('HASH(0xaaaae1660610)', 'HASH(0xaaaae17fc408)') called at /usr/local/lib/perl5/site_perl/5.36.1/JSON/Color.pm line 209
    	JSON::Color::encode_json('HASH(0xaaaae1660610)') called at bin/ai.pl line 312
    	main::api_call at bin/ai.pl line 190%
```

**Comment by perlancar** [comment]
Hi,

Can't reproduce here.

    % PERL5OPT=-Mlib::disallow=Term::Detect::Software perl -MJSON::Color=encode_json -E'say encode_json([1, "two", {three => 4}])'

(Also tested after I uninstalled `Term::Detect::Software`). `Term::Detect::Software` is an optional dependency and Color::ANSI::Util has a failover mechanism.

Which versions of the modules are you using?

**Comment by jackdeguest** [comment]
> Which versions of the modules are you using?

The latest one on CPAN, so `JSON::Color` version 0.134 and `Color::ANSI::Util` version 0.164

**Comment by perlancar** [comment]
Can you write a small test script to reproduce the problem?

**Comment by jackdeguest** [comment]
> Can you write a small test script to reproduce the problem?

I think at line 430 of `Color::ANSI::Util` you need to insert `local $@` so it does not pollute the general space.

**Comment by jackdeguest** [comment]
I have, in my script, a `local $SIG{__DIE__}` and somehow, it catches the die within your module.
By inserting `local $SIG{__DIE__}` at line 430 in `Color::ANSI::Util`, it solved the problem. Adding `local $@` is important too.

**Comment by perlancar** [comment]
Does a patch similar to this:

    diff --git a/lib/Color/ANSI/Util.pm b/lib/Color/ANSI/Util.pm
    index bc9b6b7..cac32a1 100644
    --- a/lib/Color/ANSI/Util.pm
    +++ b/lib/Color/ANSI/Util.pm
    @@ -426,6 +426,7 @@ sub _color_depth {
                     last;
                 }
                 if ($_use_termdetsw) {
    +                local $@;
                     eval { require Term::Detect::Software };
                     if (!$@) {
                         $_color_depth = Term::Detect::Software::detect_terminal_cached()->{color_depth};

fix your issue?

**Comment by perlancar** [comment]
> I have, in my script, a `local $SIG{__DIE__}` and somehow, it catches the die within your module. By inserting `local $SIG{__DIE__}` at line 430 in `Color::ANSI::Util`, it solved the problem. Adding `local $@` is important too.

Ah, I see.

How about using `Try::Tiny`?

    diff --git a/lib/Color/ANSI/Util.pm b/lib/Color/ANSI/Util.pm
    index bc9b6b7..7a4edb3 100644
    --- a/lib/Color/ANSI/Util.pm
    +++ b/lib/Color/ANSI/Util.pm
    @@ -6,6 +6,7 @@ use warnings;

     use Color::RGB::Util qw(rgb_diff);
     use Exporter qw(import);
    +use Try::Tiny;

     # AUTHORITY
     # DATE
    @@ -426,11 +427,11 @@ sub _color_depth {
                     last;
                 }
                 if ($_use_termdetsw) {
    -                eval { require Term::Detect::Software };
    -                if (!$@) {
    +                try {
    +                    require Term::Detect::Software;
                         $_color_depth = Term::Detect::Software::detect_terminal_cached()->{color_depth};
    -                    last;
    -                }
    +                    goto L1;
    +                };
                 }
                 # simple heuristic
                 if ($ENV{KONSOLE_DBUS_SERVICE}) {
    @@ -441,6 +442,7 @@ sub _color_depth {
                 $_color_depth = 16;
             }
         };
    +  L1:
         $_color_depth;
     }

**Comment by jackdeguest** [comment]
`Try::Tiny` is fine, but you will add a dependency, and you already are using `eval`. I am not sure how `Try::Tiny` works and if it localise the die signal.

**Comment by perlancar** [comment]
`Try::Tiny` will replace the `eval`, but yeah, in my opinion Try::Tiny is not tiny enough (it brings `Carp`, `Exporter`, `Sub::Util`, `constant`, `overloading`) in addition to being an XS module. Wondering if I should do the this incantation from now on everytime I want to do try-catch using eval, since I do this in various codes:

    { local $@; local $SIG{__DIE__}; eval { do_something() } if ($@) { ... } }

**Comment by jackdeguest** [comment]
> Wondering if I should do the this incantation from now on everytime I want to do try-catch using eval, since I do this in various codes:

You can always use [Nice::Try](https://metacpan.org/pod/Nice::Try), it is much more lightweight and I created it and it is closer to the real thing too :)

**Comment by perlancar** [comment]
> > Wondering if I should do the this incantation from now on everytime I want to do try-catch using eval, since I do this in various codes:
> 
> You can always use [Nice::Try](https://metacpan.org/pod/Nice::Try), it is much more lightweight and I created it and it is closer to the real thing too :)

Ah, will check Nice::Try out, thanks. `lcpan related mods Try::Tiny` didn't come up with your module :)

    % lcpan related-mods Try::Tiny
    +----------------------+-----------------------------------------------------------------------+---------------------+-----------+
    | module               | abstract                                                              | dist                | author    |
    +----------------------+-----------------------------------------------------------------------+---------------------+-----------+
    | TryCatch             | first class try catch semantics for Perl, without source filters.     | TryCatch            | ASH       |
    | Scalar::Util         | A selection of general-utility scalar subroutines                     | Scalar-List-Utils   | PEVANS    |
    | Safe::Isa            | Call isa, can, does and DOES safely on things that may not be objects | Safe-Isa            | ETHER     |
    | Throwable::Error     | an easy-to-use class for error objects                                | Throwable           | RJBS      |
    | Exception::Class     | A module that allows you to declare real exception classes in Perl    | Exception-Class     | DROLSKY   |
    | Moo                  | Minimalist Object Orientation (with Moose compatibility)              | Moo                 | HAARG     |
    | Carp                 | alternative warn and die for modules                                  | Carp                | XSAWYERX  |
    | Throwable            | a role for classes that can be thrown                                 | Throwable           | RJBS      |
    | Test::Fatal          | incredibly simple helpers for testing code with exceptions            | Test-Fatal          | RJBS      |
    | Try::Tiny::Retry     | Extends Try::Tiny to allow retries                                    | Try-Tiny-Retry      | DAGOLDEN  |
    | Moose                | A postmodern object system for Perl 5                                 | Moose               | ETHER     |
    | Syntax::Keyword::Try | a C<try/catch/finally> syntax for perl                                | Syntax-Keyword-Try  | PEVANS    |
    | Devel::StackTrace    | An object representing a stack trace                                  | Devel-StackTrace    | DROLSKY   |
    | Test::Pod::Coverage  | Check for pod coverage in your distribution.                          | Test-Pod-Coverage   | NEILB     |
    | MongoDB::Error       | MongoDB Driver Error classes                                          | MongoDB             | MONGODB   |
    | Unexpected           |                                                                       | Unexpected          | PJFL      |
    | namespace::autoclean | Keep imports out of your namespace                                    | namespace-autoclean | ETHER     |
    | Test::Exception      | Test exception-based code                                             | Test-Exception      | EXODIST   |
    | MooseX::Declare      | (DEPRECATED) Declarative syntax for Moose                             | MooseX-Declare      | ETHER     |
    | File::Slurp          | Simple and Efficient Reading/Writing/Modifying of Complete Files      | File-Slurp          | CAPOEIRAB |
    +----------------------+-----------------------------------------------------------------------+---------------------+-----------+

**Comment by perlancar** [comment]
> > Wondering if I should do the this incantation from now on everytime I want to do try-catch using eval, since I do this in various codes:
> 
> You can always use [Nice::Try](https://metacpan.org/pod/Nice::Try), it is much more lightweight and I created it and it is closer to the real thing too :)

Are you sure it's more lightweight though, it uses `PPI` :)

**Comment by jackdeguest** [comment]
> Are you sure it's more lightweight though, it uses `PPI` :)

Yes, now that you mention it. I had forgotten about it :)

**Comment by perlancar** [comment]
OK, I've committed 10a4dd6 and released 0.165.

**Comment by jackdeguest** [comment]
Thank you kindly for your swift reaction.
